appcell-resources/js/message.js

```diff
@@ -0,0 +1,474 @@
+
+    ms.getExtCellList().done(function (res) {
+        $('#extCellList').empty();
+        var results = res.d.results;
+        results.sort(function (val1, val2) {
+            return (val1.Url < val2.Url ? 1 : -1);
+        })
+        for (var i = 0; i < results.length; i++) {
+            var extCell = results[i];
+            // Displace the cell URL with the unit's proper URL. However, when sending to the server, we use "personium-localunit:" URL format.
+            var extCellUrl = ut.changeLocalUnitToUnitUrl(extCell.Url);
+            $("#extCellList").append('<option value="' + extCellUrl + '"></option>');
+        }
+    })
+
+    $("#b-create-message-sent").on('click', function () {
+        var msg = {};
+        msg.To = $("#txtDestCellUrl").val();
+        msg.Title = $("#txtMessageSubject").val();
+        msg.Body = $("#txtMessageBody").val();
+        if (ms.replyTo) {
+            msg.InReplyTo = ms.replyTo;
+        }
+        ms.postSendAPI(msg).done(function () {
+            $("#dashboard").prepend('<div class="alert alert-success alert-dismissable fade in"><a href="JavaScript:void(0)" class="close" data-dismiss="alert" aria-label="close">&times;</a><span data-i18n="message:successSendMessage"></span></div>').localize();
+            $("#modal-create-message").modal("hide");
+        }).fail(function () {
+            $("#dashboard").prepend('<div class="alert alert-warning alert-dismissable fade in"><a href="JavaScript:void(0)" class="close" data-dismiss="alert" aria-label="close">&times;</a><span data-i18n="message:failSendMessage"></span></div>').localize();
+            $("#modal-create-message").modal("hide");
+        });
+    });
+}
+
+ms.checkCellUrl = function (obj) {
+    if (obj.value) {
+        if (cm.validateCellURL(obj.value, "popupSendCellUrlErrorMsg")
+            && cm.doesUrlContainSlash(obj.value, "popupSendCellUrlErrorMsg", i18next.t("errorValidateEndWithCell"))) {
+
+            ms.getCell(obj.value).done(function () {
+                cm.getProfile(obj.value).done(function (result) {
+                    $("#txtDestProfName").val(result.DisplayName);
+                }).fail(function () {
+                    $("#txtDestProfName").val("");
+                }).always(function () {
+                    $("#popupSendCellUrlErrorMsg").html("");
+                    $("#b-create-message-sent").prop('disabled', false);
+                });
+            }).fail(function () {
+                $("#txtDestProfName").val("");
+                $("#popupSendCellUrlErrorMsg").attr("data-i18n", "notExistTargetCell").localize();
+                $("#b-create-message-sent").prop('disabled', true);
+            })
+
+        }
+
+    } else {
+        $("#txtDestProfName").val("");
+        $("#b-create-message-sent").prop('disabled', true);
+    }
+}
+
+ms.setTitle = function () {
+    cm.setTitleMenu("Message");
+}
+
+/*
+ * Display Receive Message List
+ */
+ms.openReceiveMsg = function () {
```

Comment:
This function's name: openReceivedMsgPanel

Refactor function: displayReceivedMsgList